src/librustc_typeck/check/mod.rs

```diff
@@ -3709,7 +3709,7 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
                             self.consider_hint_about_removing_semicolon(blk, expected_ty, err);
                         }
                         if let Some(fn_span) = fn_span {
-                            err.span_label(fn_span, "this function's body doesn't return");
+                            err.span_label(fn_span, "this function's body doesn't return a value");
```

Comment:
This is a bit of half-truth I don't know if we should be spreading.
All converging functions in Rust return a value, this one just returns `()`.

Reply:
The way I read this originally was "this function's body doesn't `return` a value", which is slightly more correct ^_^

Reply:
Maybe disambiguate the reading with *"this function's body doesn't `return <value>`"*?

Reply:
I don't want to be too prescriptive towards having to use `return` in case we suggest removing a semicolon

Reply:
What's the situation you are concerned about re. removing a semicolon?

Reply:
I'm just worried that by hinting at `return <value>;` we might confuse people when telling them that removing a semicolon would fix the problem. Then again it might be that it would help to cement the idea that both tail expressions and `return` statements are functionally the same... I wonder if adding a note would stating the existence of both `return` statements and tail expressions would be a good idea for these.

```rust
fn foo() -> usize {
    42;
}
```

```
error[E0308]: mismatched types
 --> src/lib.rs:1:13
  |
1 | fn foo() -> usize {
  |    ---      ^^^^^ expected usize, found ()
  |    |
  |    this function's body doesn't `return` a value
2 |     42;
  |       - help: consider removing this semicolon
  |
  = note: expected type `usize`
             found type `()`
```
vs
```
error[E0308]: mismatched types
 --> src/lib.rs:1:13
  |
1 | fn foo() -> usize {
  |    ---      ^^^^^ expected usize, found ()
  |    |
  |    this function's body doesn't `return <value>;`
2 |     42;
  |       - help: consider removing this semicolon
  |
  = note: expected type `usize`
             found type `()`
```

Reply:
Hmm... Maybe be more direct and point out that the lack of a tail expression results in `()` as a type?

```rust
fn foo() -> usize {
    42;
}
```

==>

```
rror[E0308]: mismatched types
 --> src/lib.rs:1:13
  |
1 | fn foo() -> usize {
  |    ---      ^^^^^ expected usize, found ()
  |    |
  |    this function implicitly returns `()` as the body has no tail expression
2 |     42;
  |       - help: consider removing this semicolon
  |
  = note: expected type `usize`
             found type `()`

if teaching_mode {

  = help: a function can return a value in three ways:
          - with an implicit tail expression
          - with an explicit `return <value>;`
          - where function body doesn't end with a tail expression (return type is `()`)

}
```

Reply:
That's not that bad, although I would add a mention to `return` as well:

    this function implicitly returns `()` as the body has no tail expression nor a `return` statement

Reply:
Sure; It's a tad long but it is complete :)

Reply:
changed